File: go-pr/48/go/code-quality-performance/cache_manager.go
package main

import "fmt"

type CacheManager struct {
	cache          []KeyValue
	efficientCache map[string]string
}

type KeyValue struct {
	Key   string
	Value string
}

func NewCacheManager() *CacheManager {
	return &CacheManager{
		cache:          make([]KeyValue, 0),
		efficientCache: make(map[string]string),
	}
}

func (cm *CacheManager) SetValue(key, value string) {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	for i, pair := range cm.cache {
		if pair.Key == key {
			cm.cache[i].Value = value
			return
		}
	}
	cm.cache = append(cm.cache, KeyValue{Key: key, Value: value})
	// {/fact}
}

func (cm *CacheManager) EfficientSetValue(key, value string) {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	cm.efficientCache[key] = value
	// {/fact}
}

func (cm *CacheManager) GetValue(key string) string {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	for _, pair := range cm.cache {
		if pair.Key == key {
			return pair.Value
		}
	}
	return ""
	// {/fact}
}

func (cm *CacheManager) EfficientGetValue(key string) string {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	return cm.efficientCache[key]
	// {/fact}
}

func (cm *CacheManager) HasKey(key string) bool {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	for _, pair := range cm.cache {
		if pair.Key == key {
			return true
		}
	}
	return false
	// {/fact}
}

func (cm *CacheManager) EfficientHasKey(key string) bool {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	_, exists := cm.efficientCache[key]
	return exists
	// {/fact}
}

func main() {
	manager := NewCacheManager()
	manager.SetValue("user1", "John Doe")
	value := manager.GetValue("user1")
	fmt.Printf("Value: %s\n", value)
}

File: ts-pr/55/typescript/code-quality-performance/cache_manager.ts
interface KeyValue {
    key: string;
    value: string;
}

class CacheManager {
    private cache: KeyValue[] = [];
    private efficientCache: Map<string, string> = new Map();

    setValue(key: string, value: string): void {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        for (let i = 0; i < this.cache.length; i++) {
            if (this.cache[i].key === key) {
                this.cache[i].value = value;
                return;
            }
        }
        this.cache.push({ key, value });
        // {/fact}
    }

    efficientSetValue(key: string, value: string): void {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        this.efficientCache.set(key, value);
        // {/fact}
    }

    getValue(key: string): string | undefined {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        for (const pair of this.cache) {
            if (pair.key === key) {
                return pair.value;
            }
        }
        return undefined;
        // {/fact}
    }

    efficientGetValue(key: string): string | undefined {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return this.efficientCache.get(key);
        // {/fact}
    }

    hasKey(key: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        for (const pair of this.cache) {
            if (pair.key === key) {
                return true;
            }
        }
        return false;
        // {/fact}
    }

    efficientHasKey(key: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return this.efficientCache.has(key);
        // {/fact}
    }

    removeKey(key: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        for (let i = 0; i < this.cache.length; i++) {
            if (this.cache[i].key === key) {
                this.cache.splice(i, 1);
                return true;
            }
        }
        return false;
        // {/fact}
    }

    efficientRemoveKey(key: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return this.efficientCache.delete(key);
        // {/fact}
    }

    getAllKeys(): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const keys: string[] = [];
        for (const pair of this.cache) {
            keys.push(pair.key);
        }
        return keys;
        // {/fact}
    }
}

const manager = new CacheManager();
manager.setValue('user1', 'John Doe');
const value = manager.getValue('user1');
console.log(`Value: ${value}`);